### ISSUE TITLE
openstack module (e.g. os_server  / os_network) ignore endpoint_type:

### ISSUE THREAD

**Issue by Thingee** [ansible/ansible#19611]
_From @torbatux on September 21, 2015 12:44_

##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
os_server module

##### ANSIBLE VERSION
```
[torba@localhost ~]$ ansible --version
ansible 2.0.0
```

##### SUMMARY
Hi I try to use ansible 2 with openstack support, to spin up new instances (I tried it now with different OS/installs to confirm): 

[torba@localhost ~]$ ansible --version
ansible 2.0.0

on a Centos 7 machine:
[torba@localhost ~]$ uname -r
3.10.0-229.11.1.el7.x86_64

when I try to use os_server with endpoint_type: internal and auto_floating_ip: yes
ansible still try to use the public IP (which I replaced down below with: **_PUBLIC-IP**_). 
I also tried inside of the sourced .novarc to use export OS_ENDPOINT_TYPE=internalURL which gets ignored as well.

-->SNIPP<---

```
- name: Spin up Nodes
  hosts: localhost
  connection: local
  gather_facts: false
  tasks:
    - name: Ensure host is there
      os_server:
        state: present
        endpoint_type: internal
        timeout: 100
        api_timeout: 15
        name: tbade-2
        image: 32c2c8ff-824f-4533-99ba-39d1c8d43f62
        key_name: cloud_key
        flavor: m1.medium
        region_name: RegionOne
        availability_zone: nova
        security_groups: default
        nics:
          - net-id: c79d9116-a815-4919-abcc-d17c01df3676
        auto_floating_ip: yes
        meta:
          hostname: torba-1
          group: test-torba1
      register: rax
```

-->SNAPP<---

RESULT: 

```
fatal: [localhost]: FAILED! => {"changed": false, "extra_data": null, "failed": true, "msg": "unable to get available floating IPs: Error fetching network list: Unable to establish connection to http://***PUBLIC-IP***:9696/v2.0/networks.json"}
```

So I tried then to use a minimalistic approach os_network 

-->SNIPP<---

```
- name: Create test-net
  hosts: localhost
  connection: local
  gather_facts: false
  tasks:
    - os_network:
        name: testnetwork
        state: present
```

-->SNAPP<---

RESULT:

```
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Error fetching network list: Unable to establish connection to http://***PUBLIC-IP*** :9696/v2.0/networks.json"}
```

_Copied from original issue: ansible/ansible-modules-core#2118_

**Comment by Thingee** [comment]
_From @bcoca on September 21, 2015 13:4_

cc @emonty @j2sol @Shrews

**Comment by Thingee** [comment]
_From @Shrews on September 21, 2015 20:15_

By any chance are you using a Rackspace cloud?  If so, you'll need to define a clouds.yaml in $HOME/.config/openstack and define your cloud in there. The libraries used by the modules attempts to hide cloud vendor differences (like Rackspace exposing a Neutron endpoint that isn't actually usable). The clouds.yaml helps the library to figure out what needs to be done for many (not all) cloud providers. Here is an example clouds.yaml with a Rax profile:

clouds:
    rax:
      profile: rackspace
      region_name: IAD
      auth:
        username: MY_USERNAME
        password: MY_PASSWORD
        project_name: PROJECT

Please try that and let us know if that makes things work for you.

**Comment by Thingee** [comment]
_From @Shrews on September 21, 2015 20:16_

Another attempt at the example with proper spacing:

<pre>
  clouds:
      rax:
        profile: rackspace
        region_name: IAD
        auth:
          username: MY_USERNAME
          password: MY_PASSWORD
          project_name: PROJECT
</pre>

**Comment by Thingee** [comment]
_From @emonty on September 21, 2015 20:21_

Also - let me clarify two things:

a) endpoint_type is never going to have any effect on the IP address returned from a server. That's a whole other thing, and auto_ip probably isn't what you want if you want to specify a private network to talk to the instance on

but

b) there were several bugs related to endpoint_type/interface in the underlying libraries that we just recently fixed. We'll be doing a new release with those fixes soon, so we'll ping back here soon to see if that helps you with the os_network pinging the public endpoint instead.

**Comment by Thingee** [comment]
_From @torbatux on September 21, 2015 21:30_

Sorry to say @Shrews it is RPC so private  cloud, thats also the reason, why I need the internal IP instead of the public (which is blocked by the firewall). With Rackspace public it is working with pyrax well.

So I created the .novarc (instead of $HOME/.config/openstack) to also test python novaclient

```
# COMMON CINDER ENVS
export CINDER_ENDPOINT_TYPE=internalURL

# COMMON NOVA ENVS
export NOVA_ENDPOINT_TYPE=internalURL

# test if this version support neutron env
export NEUTRON_ENDPOINT_TYPE=internalURL

# COMMON OPENSTACK ENVS
export OS_ENDPOINT_TYPE=internalURL
#export OS_PASSWORD=
export OS_AUTH_URL=http://172.29.238.10:5000/v2.0
export OS_NO_CACHE=1
export OS_TENANT_ID=b61296cb4a344c878227c06fccfc80a6
export OS_TENANT_NAME="Test"
export OS_REGION_NAME="RegionOne"

export OS_USERNAME="torba"

# With Keystone you pass the keystone password.
echo "Please enter your OpenStack Password: "
read -sr OS_PASSWORD_INPUT
export OS_PASSWORD=$OS_PASSWORD_INPUT
```

@emonty  I need the auto_ip (so floating) because the private range is not accessible by vpn ... to provision the systems I will have 512 ip of a floating range, with which I can access the created instances (or I have to do it with a jumphost).

**Comment by Thingee** [comment]
_From @emonty on September 21, 2015 21:34_

Ok. Cool - that's great to know. I'll see what we can do to support you there. Thanks for the details!

**Comment by Thingee** [comment]
_From @torbatux on September 21, 2015 21:51_

What ever I can do to support you ... let me know  ... Many thanks in advance for your effords!

**Comment by Thingee** [comment]
_From @tweisshaar on September 22, 2015 17:32_

Hi guys,

the issue is that the `shade/__init__.py` is calling the http constructor of `neutronclient/v2.0/client.py` (line   169) which itself is calling `neutronclient/client.py` without transmitting the endpoint_type.

Further the related function in `neutronclient/client.py` is assuming endpoint_type='publicURL' as default.

As I noticed that the `shade/__init__.py` is providing the `self.default_interface` with the correct setting, I assume that patch works:

`(rackspace) ~/D/t/i/dev (master) $ cat shade_init.patch`

``` patch
--- /Users/tw/.virtualenvs/rackspace/lib/python2.7/site-packages/shade/__init__.py      2015-09-22 19:13:46.000000000 +0200
+++ __init__.py 2015-09-22 19:14:54.000000000 +0200
@@ -707,8 +707,9 @@
                 token=self.auth_token,
                 session=self.keystone_session,
                 region_name=self.region_name,
-                timeout=self.api_timeout)
-        return self._neutron_client
+                timeout=self.api_timeout,
+                endpoint_type=self.default_interface)       # neutronclient/client.py uses 'publicURL' as default
+        return self._neutron_client                         # endpoint_type, as the endpoint_type is not part of kwargs!

     @property
     def designate_client(self):
```

Looking for feedback.

Cheers,
Thomas

**Comment by Thingee** [comment]
_From @Shrews on September 24, 2015 12:11_

Please check the latest version of shade. We've made many changes to how the clients are created, and I believe this problem is now solved for you, but I think we'll need a new release before you get the fix. That should be coming RSN.

https://github.com/openstack-infra/shade

**Comment by Thingee** [comment]
_From @Shrews on September 24, 2015 12:13_

Meant to add, if you could test the new version, that would be helpful.  :)

**Comment by Thingee** [comment]
_From @Shrews on September 24, 2015 12:29_

Wow, I shouldn't respond before 2nd cup of coffee.

The latest changes I looked at won't fix all the issues you pointed out.  _sigh_

**Comment by Thingee** [comment]
_From @Nosmoht on September 28, 2015 14:47_

Hi All,

i'm using shade 0.14.0 (don't know if this is the last version) and still get the error.

**Comment by Thingee** [comment]
_From @emonty on September 28, 2015 14:48_

On 09/28/2015 09:47 AM, Thomas Krahn wrote:

> Hi All,
> 
> i'm using shade 0.14.0 (don't know if this is the last version) and
> still get the error.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible-modules-core/issues/2118#issuecomment-143764499.

Blast.

**Comment by Thingee** [comment]
_From @dagnello on October 14, 2015 22:20_

I am also seeing the same issue with Shade 0.10.0.  Which version has this been resolved in?

**Comment by Thingee** [comment]
_From @ansibot on August 1, 2016 2:34_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by Thingee** [comment]
_From @ansibot on September 12, 2016 19:36_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Thingee** [comment]
_From @ansibot on October 6, 2016 15:50_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Thingee** [comment]
_From @ansibot on October 22, 2016 9:19_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Thingee** [comment]
_From @indispeq on November 3, 2016 20:12_

Hi, I am having the same issue with `shade==1.12.1`  . I am using Rackspace Private Cloud with openstack Kilo and have OS_ENDPOINT_TYPE = internalURL defined, but server creation times out with an error saying it cannot reach http://public-ip:8774/v3/xxx . I have tried looking in the code where this override happens. It seems like it is taking the PublicURL field by default and not checking if the variable is in fact defined. This works on installations where the publicurl is same as internalurl.

**Comment by Thingee** [comment]
_From @indispeq on November 3, 2016 21:0_

Just as an FYI, since it was RPC and we do not hit it from the public IP, I used the openstack client to set the endpoint URLs for public to the internal IP used by admin and internal and this way I can launch an instance. This further shows that it is indeed taking publicURL and OS_ENDPOINT_TYPE is not taken into account.

**Comment by Thingee** [comment]
_From @ansibot on November 6, 2016 10:9_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Thingee** [comment]
_From @ansibot on November 22, 2016 10:17_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, @thingee, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Thingee** [comment]
_From @ansibot on December 7, 2016 17:55_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
cc @omgjlk
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @mnaser
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mnaser** [comment]
bug_resolved

This should be fixed by the following:

https://github.com/ansible/ansible/commit/0f893027c47560777f5295a338238a820c6c954b

Please make sure to use `interface` instead or you can stick to `endpoint_type`.
